### PR TITLE
Fix Lexer EOF sentinel collision with valid Unicode code points

### PR DESCRIPTION
**Pull request by spyoungtech**
# Summary

A change [introduced](https://github.com/apple/pkl/pull/917) in 0.28.0 reimplemented the Lexer in a way that implements EOF detection using an in-band sentinel value `Short.MAX_VALUE` (0x7FFF), which collides with the valid Unicode codepoint U+7FFF. When this character appears in source code, the lexer incorrectly terminates parsing, treating all subsequent content as non-existent. 

Although this is probably a rare occurrence for typical use, this issue presents an EOF injection vulnerability, which is very similar to a null byte injection vulnerability. This vulnerability enables **steganographic configuration manipulation** with the following characteristics:

1. **Silent Failure**: No parse errors, warnings, or diagnostic output
2. **Code Review Evasion**: The attack character is visually innocuous and easily overlooked
3. **Policy Bypass**: Security controls defined after the injection point are never evaluated

This affects (at least) all versions since 0.28.0: [0.28.0, 0.28.1, 0.28.2, 0.29.0, 0.29.1]

This fix removes reliance on lookahead/sentinel comparisons to prevent early parser termination.

<details><summary><h3>Proof of Concept</h3></summary>

Consider a base network policy with subsequent hardening:

**base.pkl**:
  ```pkl
  allowedPorts = new Listing {
      80
      443
      22
  }
```

  production.pkl (legitimate):
```pkl
  amends "base.pkl"

  // Production security: remove SSH access
  allowedPorts = new Listing {
      80
      443
  }
```
  production.pkl (compromised):

```pkl
  amends "base.pkl"

  // Production security: remove port 22; SSH not allowed in China env  翿!
  allowedPorts = new Listing {
      80
      443
  }
```

 Result:
  - Expected: SSH port 22 blocked
  - Actual: SSH port 22 remains open (security bypass)

  The configuration appears syntactically valid but the hardening directive is silently ignored.

</details>

## Proposed Remediation

The fix implements out-of-band EOF signaling using a dedicated atEOF boolean flag, eliminating the sentinel value collision:

  1. EOF state tracked via explicit boolean flag rather than reserved character value
  2. All EOF detection logic migrated to flag-based checks
  3. EOF sentinel changed to Character.MAX_VALUE (0xFFFF) with proper flag handling

This ensures no valid Unicode codepoint can be misinterpreted as EOF.

### Verification

Verification is added to the existing test suite in the form of a test that validates the Unicode code points, including the boundary values (U+7FFF, U+FFFF), are correctly processed without premature termination.

## References

  - https://cwe.mitre.org/data/definitions/147.html
  - https://cwe.mitre.org/data/definitions/115.html
  - https://www.unicode.org/reports/tr55/

### CODE REVIEW

File: pkl-parser/src/main/java/org/pkl/parser/Lexer.java [hunk 1]
```diff
@@ -37,8 +37,9 @@ public class Lexer {
   private boolean isEscape = false;
   // how many newlines exist between two subsequent tokens
   protected int newLinesBetween = 0;
+  private boolean atEOF = false;
 
-  private static final char EOF = Short.MAX_VALUE;
+  private static final char EOF = Character.MAX_VALUE;
```
Comment (bioball) [inline pkl-parser/src/main/java/org/pkl/parser/Lexer.java]: This still has a conflict, no? `\uffff` is an "undefined character": https://www.compart.com/en/unicode/U+FFFF

I can still have Pkl code that looks like:

```pkl
// Hello there ￿ uffff
```

I think we can fix this by representing EOF as `-1`, and change `char lookahead` to `int lookahead`; since `-1` is never a valid character code, this eliminates any chance of conflict. And, we can go back to using `if (lookahead == EOF) in this case.

Comment (spyoungtech) [reply pkl-parser/src/main/java/org/pkl/parser/Lexer.java]: The change of the sentinel value to a noncharacter is mostly cosmetic at this point. Originally, I made the change because I thought this would be a one-line kinda thing and just that it'd be better to use a non-character [designated for internal application use](https://www.unicode.org/faq/private_use) instead of a 'real' character assigned a glyph. 

But as you point out, that change alone would just shift the problem to a different character and it could still potentially collide with other applications' legitimate use of the character and the vulnerability would remain.

However, combined with the other changes, it doesn't really matter what this sentinel is. I did add tests for `\uffff` though to ensure the case is properly covered. The primary thing is that solves the problem is that `atEOF` is never set based on the content of the source, so injecting EOF behavior is no longer possible.

The approach you describe would work. It is the approach I tried first, but it caused so many cascading behavior changes that I wasn't confident enough to go all the way down that route without understanding the possible performance/behavior impacts.

Comment (bioball) [reply pkl-parser/src/main/java/org/pkl/parser/Lexer.java]: Ah, I see. that's why `atEOF` was introduced.

In that case, I think it makes more sense to just get rid of `private static final char EOF`, because it'd just add some confusion for future maintainers.

File: pkl-parser/src/main/java/org/pkl/parser/Lexer.java [hunk 2]
```diff
@@ -98,8 +96,10 @@ private Token nextDefault() {
       sCol = col;
       ch = nextChar();
     }
-    if (ch == EOF && atEOF) {
+    if (atEOF && ch == '\uffff') {
```
Comment (bioball) [inline pkl-parser/src/main/java/org/pkl/parser/Lexer.java]:
```suggestion
    if (atEOF) {
```

File: pkl-parser/src/main/java/org/pkl/parser/Lexer.java [hunk 3]
```diff
@@ -732,7 +732,7 @@ private ParserError lexError(String msg, Span span) {
   }
 
   private ParserError unexpectedChar(char got, String didYouMean) {
-    if (got == EOF) {
+    if (got == '\uffff' && atEOF) {
```
Comment (bioball) [inline pkl-parser/src/main/java/org/pkl/parser/Lexer.java]:
```suggestion
    if (atEOF) {
```

File: pkl-parser/src/main/java/org/pkl/parser/Lexer.java [hunk 4]
```diff
@@ -690,7 +690,7 @@ private char nextChar() {
     var tmp = lookahead;
     cursor++;
     if (cursor >= size) {
-      lookahead = EOF;
+      lookahead = '\uffff';
```
Comment (bioball) [inline pkl-parser/src/main/java/org/pkl/parser/Lexer.java]:
```suggestion
```

File: pkl-parser/src/main/java/org/pkl/parser/Lexer.java [hunk 5]
```diff
@@ -39,15 +39,13 @@ public class Lexer {
...
-  private static final char EOF = Character.MAX_VALUE;
-
   public Lexer(String input) {
     source = input.toCharArray();
     size = source.length;
     if (size > 0) {
       lookahead = source[cursor];
     } else {
-      lookahead = EOF;
+      lookahead = '\uffff';
```
Comment (bioball) [inline pkl-parser/src/main/java/org/pkl/parser/Lexer.java]:
```suggestion
```
